### ISSUE TITLE
Feature request: support more flexible handling of seeds and seed-like objects like `numpy.random.SeedSequence`

### ISSUE THREAD

**Issue by gehring** [google/jax#4399]
This is definitely a very minor thing but probably a very common inconvenience, certainly has been for me. Converting a source of seed from numpy to jax has always felt a bit awkward to me. Part of it likely comes from the fact that there are many ways of doing this, some better than others, but also because of how it contrast with how seamless and easy it is to deal with RNGs in both frameworks in isolation.

Given the similarities between the `np.random.SeedSequence` API and jax's `PRNGKey` api, I feel like it would be very intuitive and convenient if the conversion from `SeedSequence` to `PRNGKey` was as flexible as how `numpy.ndarray` and `jax.DeviceArray` are handled (and the abstract counter-parts).

If that feels like too big of a leap, I would suggest expanding types accepted by `jax.random.PRNGKey` to match the types accepted by `numpy.random.default_rng`. The automatic handling of `SeedSequence` would definitely be very cool, even just the simpler idea would greatly contribute to enabling a smooth (and dummy proof) transition from numpy's random API (version >= 1.17) to jax's random API.

Here's how you could do this with single dispatching (untested draft code). You could do it with a bunch of `elif` too but where's the fun in that.
```python
import functools

import numpy as np

import jax.numpy as jnp
from jax import lax


@functools.singledispatch
def PRNGKey(seed) -> jnp.ndarray:
  """Create a pseudo-random number generator (PRNG) key given an integer seed.

  Args:
    seed: a 64- or 32-bit integer used as the value of the key.

  Returns:
    A PRNG key, which is modeled as an array of shape (2,) and dtype uint32. The
    key is constructed from a 64-bit seed by effectively bit-casting to a pair
    of uint32 values (or from a 32-bit seed by first padding out with zeros).
  """
  raise NotImplementedError


@PRNGKey.register
def _seedsequence_to_prngkey(seed: np.random.SeedSequence) -> jnp.ndarray:
  return jnp.array(seed.generate_state(2, np.uint32))


def _convert(k):
    return lax.reshape(lax.convert_element_type(k, np.uint32), [1])


@PRNGKey.register
def _int_to_prngkey(seed: int) -> jnp.ndarray:
  # This is the old logic but we could just feed the seed to SeedSequence and
  # then convert that.
  k1 = _convert(np.bitwise_and(np.right_shift(seed, 32), 0xFFFFFFFF))
  k2 = _convert(jnp.bitwise_and(seed, 0xFFFFFFFF))
  return lax.concatenate([k1, k2], 0)


@PRNGKey.register
def _ndarray_to_prngkey(seed: np.ndarray) -> jnp.array:
    return PRNGKey(np.random.SeedSequence(seed))


@PRNGKey.register
def _numpy_generator_to_prngkey(seed: np.random.Generator) -> jnp.array:
    return jnp.array(
        seed.integers(0, np.iinfo(np.uint32).max, (2,), dtype=np.uint32,
                      endpoint=True))


@PRNGKey.register
def _numpy_bit_generator_to_prngkey(seed: np.random.BitGenerator) -> jnp.array:
    return PRNGKey(np.random.default_rng(seed))
```
Note: registering with type annotation requires python >= 3.7. Otherwise you have to explicitly specify it.

I'd be happy to wrap this up in a PR if there is any interest.

**Comment by jakevdp** [comment]
Hi – this is an interesting idea, and definitely would make ineroperability between `np.random` and `jax.random` easier.

One concern I have is that numpy's random interface is stateful, where JAX's is not. This makes things behave in possibly unexpected ways when they are combined with JAX's jit. For example:
```python
from jax import jit
import jax.numpy as jnp
import numpy as np

def _numpy_generator_to_prngkey(seed: np.random.Generator) -> jnp.ndarray:
    return jnp.array(
        seed.integers(0, np.iinfo(np.uint32).max, (2,), dtype=np.uint32,
                      endpoint=True))
    
seed = np.random.default_rng(seed=42)

print(_numpy_generator_to_prngkey(seed) == _numpy_generator_to_prngkey(seed))
# False

_numpy_generator_to_prngkey = jit(_numpy_generator_to_prngkey, static_argnums=[0])
print(_numpy_generator_to_prngkey(seed) == _numpy_generator_to_prngkey(seed))
# True
```
In the jitted version of the function, the seed is only generated once at tracing time. I suspect that this sort of thing might lead to subtle bugs that would be very difficult to identify.

Do you have thoughts on that?

**Comment by gehring** [comment]
Numpy's stateful generators definitely complicates things but I think this can be handled transparently and (hopefully) intuitively.

`np.random.default_rng` supports `int`, `np.ndarray`, `np.random.SeedSequence`, `np.random.Generator`, and `np.random.BitGenerator`. For `int` and `np.ndarray`, this obviously doesn't require anything special and is already currently handled by `PRNGKey`. As for `SeedSequence`, it should be straightforward since, as I understand it, it is immutable and conceptually quite similar to `PRNGKey` so the generators are the only tricky part.

The simplest would be to simply accept it as a limitation and just disallow using generators as seeds; their API is the least compatible with PRNGKey after all. The second simplest would be to disallow this type of conversion while inside a jitted function (probably also in any transformed function) while still allowing it outside. That's what I had in mind when I opened this issue. We might be able to do better but this type of 'cautious error raising when jitting' approach would be reasonably aligned with other "preemptive" errors `jax.jit` will catch.

Otherwise, a more complex solution would be to convert a generator into a np.ndarray and then treating it similarly to how numpy's mutable arrays are handled in terms of deciding when to treat it as static and when to treat it as some. This would require some careful thought about all the different ways a user could cause this conversion to make sure it doesn't lead to anything unintuitive. This is by far the more auto-magical solution and it might not be worth the complexity, especially given the goal would be to make the transitions between numpy and jax simpler. Requiring the user to read through several paragraphs of caveats would certainly defeat the purpose...

I'm in favor of simply disallowing generators when the intent of the user might be ambiguous, wherever that complexity threshold might be. That might require the conversion rules to so strict as to forbidding generators altogether, but maybe we can allow some simple but possible common use cases. I'd have to start writing out examples to convince myself one way or the other. Alternatively, we could also make the generator conversion a buy-in kind of thing by providing some helper function for converting it explicitly.

That was a bit of stream of consciousness but hopefully that made sense! WDYT?
